### DISCUSSION TITLE
Problems with framerate

### DISCUSSION THREAD

**Discussion by Axagthoth** [sstaub/sACN#5] [Q&A] [answered]
Hi Stefan! First of all thanks a lot for this amazing library. It's easy to grasp and yet incredibly powerful, and just works out of the box! However, I'm running into problems with the framerate, probably due to my own lack of expertise, so I thought I'd ask you for help.

First warning! I'm trying to run this over WiFi, using DHCP. Hopefully that's not a big no-no to begin with? Just running the receiver example, I can never get above 16fps. I tried with an Asus AC87U router connected to the internet (other bunch of packets on that network), and with a dedicated Asus RT-N12 (just the sending laptop and the receiving Arduino Uno R4 on the LAN).
I'm sending from MadMapper and want to convert that data into FastLED stuff. Everything runs like a charm except for those FPS. I even got rid of all the FastLED code but still hit that 13-16fps limit.

While I don't know MadMapper's implementation of sACN, I'm currently just sending 192 channels on a single universe. Even if it was using 32-bit words per channel, that's 2KB of data per universe per frame, or 240KB per second at 120fps. That shouldn't be able to saturate the network right? The 48MHz of the Uno R4 shouldn't be a problem either? Have you encountered such problems before and would you be willing to share your insights? Thanks a lot in advance!

P.S.: while I don't currently own an Ethernet Shield, I could try that, but it sort of defeats the purpose of having everything nice and wireless :-|

**Comment by sstaub** [comment, marked as answer]
Sending DMX Data over WLAN is always a problem. It works only with dedicated DMX hardware like CRMX which allows a stable connection. You should use a wired Ethernet cable for this purpose. Maybe MadMapper allows sACN over Unicast instead of Multicast.

**Reply by Axagthoth** [reply]
I was afraid that might be the reason :-( While I can't try the wired connection, acting on a hunch I swapped out the board for a Nano 33 IoT (which uses the WiFiNINA library instead). Guess what? Already the frames bumped up to 25-45. I'm guessing the WiFiS3 library still needs ironing out as the R4 is a relatively new product? I'll try with an MKR1000 and an ESP8266 and report back my findings.

Allow me to use this occasion to thank you again for all of your Arduino-Lighting related work, it's amazing stuff!!!

**Reply by Axagthoth** [reply]
Ok, sorry for the delay. Here are my findings, just in case it helps someone:
- Swapping out for different WiFi architectures doesn't really do much, although at the time of writing the Arduino UNO R4 is definitely the laggiest option.
- Slapping on an Ethernet shield and wiring the connection on the board side already does wonders for performance (getting stable 40fps, no noticeable lagginess). An issue that appears in this scenario are random dips in fps, which can be naively solved by adding a delay(x) to the loop function with a low value for x (between 10 and 20 worked best for me).
- Naturally, the best option is wired connections on both sides, then it works on 100fps+ with no noticeable dips whatsoever.

**Comment by sstaub** [comment]
The frame rate depend also much on the microcontroller you use. Personal I use a Wiznet W5500-EVB-Pico. I didn't test with Teensy 4.1 but it should also give good results. Also the SPI settings can bring problems.
